### PR TITLE
Close resources via try-with-resources

### PR DESCRIPTION
**Pull request by JanStureNielsen**

### CODE REVIEW

File: applications/binaryfile-recorder/src/main/java/com/paritytrading/nassau/binaryfile/recorder/Recorder.java [hunk 1]
```diff
@@ -170,46 +165,44 @@ public void endOfSession(MoldUDP64Client session) {
...
-
-        selector.close();
+        }
     }
 
     private static void receive(SoupBinTCPClient client) throws IOException {
-        Selector selector = Selector.open();
+        try (Selector selector = Selector.open()) {
 
-        SelectionKey key = client.getChannel().register(selector, SelectionKey.OP_READ);
```
Comment (JanStureNielsen) [inline applications/binaryfile-recorder/src/main/java/com/paritytrading/nassau/binaryfile/recorder/Recorder.java]: I removed the unused local `SelectionKey key` here which probably should have been in its own PR.

Comment (jvirtanen) [reply applications/binaryfile-recorder/src/main/java/com/paritytrading/nassau/binaryfile/recorder/Recorder.java]: I think it's fine to clean up here. Nice find! 👍

File: applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Events.java [hunk 2]
```diff
@@ -76,15 +76,9 @@ public void receive() throws IOException {
...
-
-                        try {
+                        try (Receiver receiver = (Receiver)key.attachment()) {
                             receiver.receive();
-                        } catch (IOException e1) {
-                            try {
-                                receiver.close();
-                            } catch (IOException e2) {
-                            }
+                        } catch (IOException e) {
```
Comment (jvirtanen) [inline applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Events.java]: Is this doing the right thing? 🤔  `receiver.close()` should be invoked if an `IOException` occurs in the `receiver.receive()` invocation. To me, it looks like `receiver.close()` would get called unconditionally when exiting the `try` block.

File: tests/binaryfile-perf-test/src/main/java/com/paritytrading/nassau/binaryfile/perf/PerfTest.java [hunk 3]
```diff
@@ -28,17 +28,18 @@ public void endOfSession() {
 
         BinaryFILEStatusParser parser = new BinaryFILEStatusParser(counter, statusListener);
 
-        BinaryFILEReader reader = BinaryFILEReader.open(new File(inputFilename), parser);
+        double seconds;
 
-        long started = System.nanoTime();
+        try (BinaryFILEReader reader = BinaryFILEReader.open(new File(inputFilename), parser)) {
```
[the remark this answers is not in the document]
Comment (JanStureNielsen) [reply tests/binaryfile-perf-test/src/main/java/com/paritytrading/nassau/binaryfile/perf/PerfTest.java]: Makes sense; looks & feels like better.

File: applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Events.java (hunk 2 above)
Comment (JanStureNielsen) [reply applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Events.java]: Right you are -- my mistake -- undoing that change.

File: applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Events.java [hunk 4]
```diff
@@ -80,7 +80,7 @@ public void receive() throws IOException {
 
                         try {
                             receiver.receive();
-                        } catch (IOException e1) {
+                        } catch (IOException e) {
```
Comment (jvirtanen) [inline applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Events.java]: Perhaps we can clean up also this little change to keep the diff at the minimum? 🤓

File: applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Events.java [hunk 5]
```diff
@@ -106,13 +106,8 @@ public void receive() throws IOException {
...
             for (int i = 0; i < toCleanUp.size(); i++) {
-                Session session = toCleanUp.get(i);
-
-                toKeepAlive.remove(session);
-
-                try {
-                    session.close();
-                } catch (IOException e) {
+                try (Session session = toCleanUp.get(i)) {
+                    toKeepAlive.remove(session);
```
Comment (jvirtanen) [inline applications/soupbintcp-gateway/src/main/java/com/paritytrading/nassau/soupbintcp/gateway/Events.java]: This change took a while to grow on me, but now I think it's definitely better than before. 👍

**Comment by JanStureNielsen** [comment]
How about I submit a clean single-commit PR with these changes?

**Comment by jvirtanen** [comment]
A fairly easy way would be to do it like this:
```
git reset --soft 667e7f2287e273e9bedf0d7ef8f4d98661c402d4
git commit -m "Close resources via try-with-resources."
git push -f origin try-with-resources
```

**Comment by JanStureNielsen** [comment]
Branch updated to address the comments and squash, @jvirtanen.

**Comment by jvirtanen** [comment]
Great! Thank you, @JanStureNielsen! 🙌

**Pull request merged**